### PR TITLE
[CMake] Make CXX_FLAGS_<BuildType> cache variables for linux.

### PR DESCRIPTION
**Pull request by hageboeck**
Currently, all user-provided build flags are overwritten without notice.
By making the build flags cache variables, they are only set if the user
didn't set them explicitly.
Further, the printing of the build flags now also correctly prints the
flags being set by the build type.

### CODE REVIEW

File: cmake/modules/CheckCompiler.cmake [hunk 1]
```diff
@@ -195,5 +195,6 @@ int main() {}
 #---Print the final compiler flags--------------------------------------------------------------------
 message(STATUS "ROOT Platform: ${ROOT_PLATFORM}")
 message(STATUS "ROOT Architecture: ${ROOT_ARCHITECTURE}")
-message(STATUS "Build Type: ${CMAKE_BUILD_TYPE}")
-message(STATUS "Compiler Flags: ${CMAKE_CXX_FLAGS} ${CMAKE_CXX_FLAGS_${uppercase_CMAKE_BUILD_TYPE}}")
+string(TOUPPER ${CMAKE_BUILD_TYPE} uppercase_CMAKE_BUILD_TYPE)
+message(STATUS "Build Type: ${CMAKE_BUILD_TYPE} (flags = '${CMAKE_CXX_FLAGS_${uppercase_CMAKE_BUILD_TYPE}}')")
+message(STATUS "Compiler Flags: ${CMAKE_CXX_FLAGS_${uppercase_CMAKE_BUILD_TYPE}} ${CMAKE_CXX_FLAGS}")
```
Comment (amadio) [inline cmake/modules/CheckCompiler.cmake]: Is `${CMAKE_CXX_FLAGS_${uppercase_CMAKE_BUILD_TYPE}}` really coming before `${CMAKE_CXX_FLAGS}` during the build? I thought it was the reverse...

File: cmake/modules/SetUpLinux.cmake [hunk 2]
```diff
@@ -106,18 +106,18 @@ if(CMAKE_COMPILER_IS_GNUCXX)
...
+  set(CMAKE_CXX_FLAGS_OPTIMIZED      "-Ofast -DNDEBUG"  CACHE STRING "Flags for a highly optimised build")
+  set(CMAKE_CXX_FLAGS_DEBUG          "-g"               CACHE STRING "Flags for a debug build")
+  set(CMAKE_CXX_FLAGS_DEBUGFULL      "-g3"              CACHE STRING "Flags for a debug build with more debug info")
+  set(CMAKE_CXX_FLAGS_PROFILE        "-g3 -ftest-coverage -fprofile-arcs" CACHE STRING "")
+  set(CMAKE_C_FLAGS_RELWITHDEBINFO   "-O3 -g -DNDEBUG"  CACHE STRING "Flags for release build with debug info")
+  set(CMAKE_C_FLAGS_RELEASE          "-O3 -DNDEBUG"     CACHE STRING "Flags for release build")
+  set(CMAKE_C_FLAGS_OPTIMIZED        "-Ofast -DNDEBUG"  CACHE STRING "Flags for a highly optimised build")
+  set(CMAKE_C_FLAGS_DEBUG            "-g"               CACHE STRING "Flags for a debug build")
+  set(CMAKE_C_FLAGS_DEBUGFULL        "-g3 -fno-inline"  CACHE STRING "Flags for a debug build with more debug info")
+  set(CMAKE_C_FLAGS_PROFILE          "-g3 -fno-inline -ftest-coverage -fprofile-arcs" CACHE STRING "")
```
Comment (amadio) [inline cmake/modules/SetUpLinux.cmake]: @Axel-Naumann I wonder if we really need all those build types. We could move to use only the usual `Release`, `Debug`, `RelWithDebInfo`, and `MinSizeRel`, which are the defaults provided by CMake. Maybe we can leave these settings as CMake's defaults, which should be reasonable enough, and set [`CMAKE_<LANG>_FLAGS_INIT`](https://cmake.org/cmake/help/v3.9/variable/CMAKE_LANG_FLAGS_INIT.html) and friends if needed.

File: cmake/modules/CheckCompiler.cmake (hunk 1 above)
Comment (hageboeck) [reply cmake/modules/CheckCompiler.cmake]: You are right. The wrong order is from the old PR where I wanted to change the order of the compile flags.
Will be fixed soon.

File: cmake/modules/SetUpLinux.cmake (hunk 2 above)
Comment (hageboeck) [reply cmake/modules/SetUpLinux.cmake]: That's not my business. I don't know who added them in the first place ...

Comment (Axel-Naumann) [reply cmake/modules/SetUpLinux.cmake]: I don't have a strong opinion, we should do what you recommend. But that can probably be done in a subsequent PR?

Comment (amadio) [reply cmake/modules/SetUpLinux.cmake]: Yeah, go ahead, we can continue in subsequent PRs.

**Pull request merged**